src/radio/index.js

```diff
@@ -341,38 +341,49 @@ class Radio extends EventEmitter {
       this.history.unshift(this.current);
       while (this.history.length > 20) this.history.pop();
 
-      this.app.db.lpush('radio:history', JSON.stringify(this.current)); // TODO
+      this.app.db.lpush('radio:history', JSON.stringify(this.current));
     }
 
-    if (this.order.length > 0) {
-      // $FlowFixMe
-      const index = this.order.findIndex(u => this.queues.has(u.id) && this.queues.get(u.id).length > 0);
-      if (index !== -1) {
-        const user = this.order[index];
-        const queue = this.queues.get(user.id) || [];
-        const data = queue.shift();
-
-        this.order.push(...this.order.splice(0, index + 1));
-
-        const {fp} = data;
-        this.current = data.song;
-        // $FlowFixMe
-        this.current.player = {
-          dj: trimUser(user),
-          startTime: Date.now(),
-        };
-
-        this.emit('song', fp, this.current);
-        this.emit('order', this.order);
-        this.emit('queue', user, queue);
-        return;
+    let orderIndex = -1;
+    let songIndex;
+    let user;
+    let queue;
+
+    for (let i = 0; i < this.order.length; i++) {
+      user = this.order[i];
+      queue = this.queues.get(user.id);
+      if (queue) {
+        songIndex = queue.findIndex(item => item.status === QueueItemStatus.DONE);
+        if (songIndex !== -1) {
+          orderIndex = i;
+          break;
+        }
       }
     }
 
-    if (this.current) {
-      this.current = null;
-      this.emit('song', null);
+    if (orderIndex === -1) {
+      if (this.current) {
+        this.current = null;
```

Comment:
So is this for setting song order? What is this checking? If it's the last song?

Reply:
A common sentinel value for not finding something in JS is -1 (see [Array#indexOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf), for instance).

Higher up, we iterate all DJs from `order` to check who should go next. If they qualify, we set `orderIndex` to their index in `order`. If we never find someone who qualifies, it retains its original value of -1.

If we didn't find an eligible DJ and song, but we're currently playing a song, we should update internal state to reflect that we are no longer playing a song.

---

The logic and handling for this feels really bad to me, but it grew out of some refactoring from the way we have to handle things now that unfinished items show up in the queue. You can refactor this some more after merge, if you'd like.